service/src/main/java/bio/terra/workspace/service/policy/TpsRetry.java

```diff
@@ -0,0 +1,145 @@
+    return apiException.getCode() == TIMEOUT_STATUS_CODE
+        && apiException.getCause() instanceof SocketTimeoutException;
+  }
+
+  public static <T> T retry(TpsRetry.TpsFunction<T> function)
+      throws ApiException, InterruptedException {
+    TpsRetry TpsRetry = new TpsRetry();
+    return TpsRetry.perform(function);
+  }
+
+  public static <T> T retry(TpsRetry.TpsFunction<T> function, Duration timeout)
+      throws ApiException, InterruptedException {
+    TpsRetry TpsRetry = new TpsRetry(timeout);
+    return TpsRetry.perform(function);
+  }
+
+  public static void retry(TpsRetry.TpsVoidFunction function)
+      throws ApiException, InterruptedException {
+    TpsRetry TpsRetry = new TpsRetry();
+    TpsRetry.performVoid(function);
+  }
+
+  public static void retry(TpsRetry.TpsVoidFunction function, Duration timeout)
+      throws ApiException, InterruptedException {
+    TpsRetry TpsRetry = new TpsRetry(timeout);
+    TpsRetry.performVoid(function);
+  }
+
+  private <T> T perform(TpsRetry.TpsFunction<T> function)
+      throws ApiException, InterruptedException {
+    while (true) {
+      try {
+        return function.apply();
+      } catch (ApiException ex) {
+        if (isRetryable(ex)) {
+          logger.info("TpsRetry: caught retry-able exception: ", ex);
+          sleepOrTimeoutBeforeRetrying(ex);
+        } else {
+          throw ex;
+        }
+      }
+    }
+  }
+
+  private boolean isRetryable(ApiException apiException) {
+    return isTimeoutException(apiException)
+        || apiException.getCode() == HttpStatus.SC_INTERNAL_SERVER_ERROR
+        || apiException.getCode() == HttpStatus.SC_BAD_GATEWAY
+        || apiException.getCode() == HttpStatus.SC_SERVICE_UNAVAILABLE
+        || apiException.getCode() == HttpStatus.SC_GATEWAY_TIMEOUT;
+  }
+
+  private void performVoid(TpsRetry.TpsVoidFunction function)
+      throws ApiException, InterruptedException {
+    perform(
+        () -> {
+          function.apply();
+          return null;
+        });
+  }
+
+  /**
+   * Given an exception from Tps, either timeout and rethrow the error from Tps or sleep for
+   * retryDuration. If the thread times out while sleeping, throw the initial exception.
+   *
+   * @param previousException The error Tps threw
+   * @throws ApiException InterruptedException
+   */
+  private void sleepOrTimeoutBeforeRetrying(ApiException previousException)
+      throws ApiException, InterruptedException {
+    if (operationTimeout.minus(retryDuration).isBefore(now())) {
+      logger.error("TpsRetry: operation timed out after " + operationTimeout.toString());
+      // If we timed out, throw the error from Tps that caused us to need to retry.
+      throw previousException;
+    }
+    logger.info("TpsRetry: sleeping " + retryDuration.getSeconds() + " seconds");
+    TimeUnit.SECONDS.sleep(retryDuration.getSeconds());
+    retryDuration = retryDuration.plus(INITIAL_WAIT);
+    if (retryDuration.compareTo(MAXIMUM_WAIT) > 0) {
```

Comment:
nit, this can be one line: retryDuration = max(retryDuration, Maximum_WAIT). 

but also could you mention in the description that the retry duration will increase by 10 seconds each time.

Reply:
Hmm... I couldn't find a `max` that compares Duration. Do you have one handy?
I could use `a.compareTo(b) >= 0 ? a : b`, but repeats the `retryDuration.plus(INITIAL_WAIT)`.
So I have left it as is.

I added more explanation to the description.